src/Data/Tree/AVL/Indexed/Relation/Unary/Any.agda

```diff
@@ -0,0 +1,118 @@
+open import Relation.Nullary.Sum using (_⊎-dec_)
+open import Relation.Unary
+
+open StrictTotalOrder strictTotalOrder renaming (Carrier to Key)
+open import Data.Tree.AVL.Indexed strictTotalOrder as AVL
+  using (Tree; Value; Key⁺; [_]; _<⁺_; K&_; _,_; key; _∼_⊔_; ∼0; leaf; node)
+
+
+private
+  variable
+    v p q : Level
+    V : Value v
+    P : Pred (K& V) p
+    Q : Pred (K& V) q
+    l u : Key⁺
+    n : ℕ
+    t : Tree V l u n
+
+------------------------------------------------------------------------
+-- Definition
+
+-- Given a predicate P, Any P t is a path to one element in t that satisfies P.
+-- There may be others.
+-- See `Relation.Unary` for an explanation of predicates.
+
+data Any {V : Value v} (P : K& V → Set p) {l u}
+     : ∀ {n} → Tree V l u n → Set (p ⊔ a ⊔ v ⊔ ℓ₂) where
+  here  : ∀ {hˡ hʳ h} {kv : K& V} → P kv →
+          {lk : Tree V l [ kv .key ] hˡ}
+          {ku : Tree V [ kv .key ] u hʳ}
+          {bal : hˡ ∼ hʳ ⊔ h} →
+          Any P (node kv lk ku bal)
+  left  : ∀ {hˡ hʳ h} {kv : K& V}
+          {lk : Tree V l [ kv .key ] hˡ} →
+          Any P lk →
+          {ku : Tree V [ kv .key ] u hʳ}
+          {bal : hˡ ∼ hʳ ⊔ h} →
+          Any P (node kv lk ku bal)
+  right : ∀ {hˡ hʳ h} {kv : K& V}
+          {lk : Tree V l [ kv .key ] hˡ}
+          {ku : Tree V [ kv .key ] u hʳ} →
+          Any P ku →
+          {bal : hˡ ∼ hʳ ⊔ h} →
+          Any P (node kv lk ku bal)
+
+------------------------------------------------------------------------
+-- Operations on Any
+
+map : P ⊆ Q → Any P t → Any Q t
+map f (here  p) = here (f p)
+map f (left  p) = left (map f p)
+map f (right p) = right (map f p)
+
+lookup : {t : Tree V l u n} → Any P t → Key
+lookup (here {kv = kv} _) = kv .key
+lookup (left  p)          = lookup p
+lookup (right p)          = lookup p
+
+-- If any element satisfies P, then P is satisfied.
+
+satisfied : Any P t → ∃ P
+satisfied (here  p) = -, p
+satisfied (left  p) = satisfied p
+satisfied (right p) = satisfied p
+
+module _ {hˡ hʳ h} {kv : K& V}
+  {lk : Tree V l [ kv .key ] hˡ}
+  {ku : Tree V [ kv .key ] u hʳ}
+  {bal : hˡ ∼ hʳ ⊔ h}
+  where
+
+  toSum : Any P (node kv lk ku bal) →
+          P kv ⊎ Any P lk ⊎ Any P ku
+  toSum (here p)  = inj₁ p
+  toSum (left p)  = inj₂ (inj₁ p)
+  toSum (right p) = inj₂ (inj₂ p)
+
+  fromSum : P kv ⊎ Any P lk ⊎ Any P ku →
+            Any P (node kv lk ku bal)
+  fromSum (inj₁ p)        = here p
+  fromSum (inj₂ (inj₁ p)) = left p
+  fromSum (inj₂ (inj₂ p)) = right p
+
+------------------------------------------------------------------------
+-- Properties of predicates preserved by Any
+
+any? : Decidable P → (t : Tree V l u n) → Dec (Any P t)
+any? P? (leaf _)          = no λ ()
+any? P? (node kv l r bal) = map′ fromSum toSum
+  (P? kv ⊎-dec any? P? l ⊎-dec any? P? r)
+
+satisfiable : ∀ {k l u} → l <⁺ [ k ] → [ k ] <⁺ u →
+              Satisfiable (P ∘ (k ,_)) →
+              Satisfiable {A = Tree V l u 1} (Any P)
```

Comment:
Why are we not quantifying over the height here even though we are quantifying over the height in `any?`?

Reply:
It's not clear how to populate the subtrees. I guess we could
take subtrees as arguments too but then this function is basically
just `node` so I'm not convinced it's of any use.

The `List` equivalent generates a singleton list rather than taking
an arbitrary tail for instance.